File: JavaScript/501-Find Mode in Binary Search Tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[]}
 */
var findMode = function(root) {
    let res = new Array(), stack = new Array();
    let preNode = null, count = 0, maxCount = 0;
    
    
    while (true){
        while (root){
            stack.push(root);
            root = root.left;
        }
        if (stack.length){
            root = stack.pop();
            if (preNode && root.val !== preNode.val){
                if (count > maxCount){
                    res = [preNode.val];
                    maxCount = count;
                }else if(count === maxCount){
                    res.push(preNode.val)
                }
                count = 1;
            }else{
                count += 1;
            }
            preNode = root;
            root = root.right;
        }else{
            break;
        }
    }
    
    if (count > maxCount){
        res = [preNode.val];
    }else if(count === maxCount){
        res.push(preNode.val);
    }
    
    return res;
};

File: C++/501-Find Mode in Binary Search Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
    vector<int> res;
    int mx = 0;
    TreeNode* pre = nullptr;
    int count = 0;
    void dfs(TreeNode* node){
        if (node == nullptr) return;
        dfs(node->left);
        if (pre != nullptr && pre->val == node->val){
            count++;
        } else{
            count = 1;
        }
        if (count > mx){
            mx = count;
            res.clear();
            res.push_back(node->val);
        } else if (count == mx){
            res.push_back(node->val);
        }
        pre = node;

        dfs(node->right);
    }
    
public:
    vector<int> findMode(TreeNode* root) {
        dfs(root);
        return res;
    }
};